yarn-project/world-state/src/native/world_state_ops_queue.ts

```diff
@@ -0,0 +1,192 @@
+      this.requests.push(op);
+    }
+  }
+
+  // Non mutating requests including uncommitted state
+  private executeNonMutatingUncommitted(op: WorldStateOp) {
+    // If there are no mutating requests in flight and there is nothing queued
+    // then send the request immediately
+    // If a mutating request is in flight then we must wait
+    // If a mutating request is not in flight but something is queued then it must be a mutating request
+    if (this.inFlightMutatingCount == 0 && this.requests.length == 0) {
+      this.sendEnqueuedRequest(op);
+    } else {
+      this.requests.push(op);
+    }
+  }
+
+  private executeNonMutatingCommitted(op: WorldStateOp) {
+    // This is a non-mutating request for committed data
+    // It can always be sent
+    op.request()
+      .then(resp => {
+        op.promise.resolve(resp);
+        this.ops.delete(op.requestId);
+      })
+      .catch(err => {
+        op.promise.reject(err);
+        this.ops.delete(op.requestId);
+      });
+  }
+
+  private checkAndEnqueue(completedOp: WorldStateOp) {
+    // As request has completed
+    // First we decrements the relevant in flight counters
+    if (completedOp.mutating) {
+      --this.inFlightMutatingCount;
+    }
+    --this.inFlightCount;
+
+    // If there are still requests in flight then do nothing further
+    if (this.inFlightCount != 0) {
+      return;
+    }
+
+    // No requests in flight, send next queued requests
+    // We loop and send:
+    // 1 mutating request if it is next in the queue
+    // As many non-mutating requests as we encounter until
+    // we exhaust the queue or we reach a mutating request
+    while (this.requests.length > 0) {
+      const next = this.requests[0];
+      if (next.mutating) {
+        if (this.inFlightCount == 0) {
+          // send the mutating request
+          this.requests.splice(0, 1);
+          this.sendEnqueuedRequest(next);
+        }
+        // this request is mutating, we need to stop here
+        break;
+      } else {
+        // not mutating, send and go round again
+        this.requests.splice(0, 1);
+        this.sendEnqueuedRequest(next);
+      }
+    }
+
+    // If the queue is empty, there is nothing in flight and we have been told to stop, then resolve the stop promise
+    if (this.inFlightCount == 0 && this.stopResolve !== undefined) {
+      this.stopResolve();
+    }
+  }
+
+  private sendEnqueuedRequest(op: WorldStateOp) {
+    // Here we increment the in flight counts before sending
+    ++this.inFlightCount;
+    if (op.mutating) {
+      ++this.inFlightMutatingCount;
+    }
+
+    // Make the request and pass the response/error through to the stored promise
+    op.request()
+      .then(resp => {
+        op.promise.resolve(resp);
+        this.checkAndEnqueue(op);
+        this.ops.delete(op.requestId);
+      })
+      .catch(err => {
+        op.promise.reject(err);
+        this.checkAndEnqueue(op);
+        this.ops.delete(op.requestId);
+      });
```

Comment:
Same as above. If the ops.delete was moved inside `checkAndEnqueue` it could be even cleaner I think

```suggestion
    op.request().then(op.promise.resolve, op.promise.reject)
      .finally(resp => {
        this.checkAndEnqueue(op);
        this.ops.delete(op.requestId);
      })
```